### ISSUE TITLE
Making Vuex-cloud compatible with other stores

### ISSUE THREAD

**Issue by louisameline** [vuex-cloud/rfcs#15]
What do we have to keep in mind to be able to plug the lib on another store?

The core concept is that it's Vuex-setters which will handle the vuex-specific work.
Basically, Vuex-setters will trigger events to say "the user wants this data", and Vuex-cloud-firestore will call a method on Vuex-setters to say "here it is", but won't insert it in the store by itself. It doesn't need to know how the store is structured, or which methods to use. The logic of store updates will be left to Vuex-setters which will act as a connector.

It means that Vuex-cloud-firestore should not call `dispatch` or `commit` by itself, it will listen to / call the Vuex-setters API which will in turn do what's necessary on the store object. Tomorrow it could be plugged on some "VueY-setters" or "VueZ-setters" or whatever which will share the same API but for other stores.

Now the big question: should the name of our organisation be Vue-cloud instead, to be future-proof? This point is discussed in https://github.com/vuex-cloud/rfcs/issues/1

**Comment by mesqueeb** [comment]
I think it's totally OK and possible to make it compatible with other stores, and as you said, it can be done via an event listener and dispatcher.

So let's discuss the promotion and naming in the other thread and the core design here.

**Comment by mesqueeb** [comment]
Our core logic could be just a function or a class instance

I'm thinking this architecture seems easiest for us to work on:

```
[ vc/firestore ]
  - provides: read functions; write logic

[ vc/core ]
  - provides: listeners; events dispatchers; hooks

[ vc/vuex-setters ]
  - provides: write actions; logic to write to vuex store
```

So any action goes through core which takes care of passing it on & executing hooks etc.

**Comment by mesqueeb** [comment]
I spent about 3 hours thinking further about this core design and how it could be setup and created so all parts can talk to each other. And I think I've got a solid proposition at #11

**Comment by nothingismagick** [comment]
I came here after the naming convention discussion, but my VERY first thought was, awesome! Now I can sync offline. Then I saw the heavy cloud focus and felt a little sad. To this end, and to overcome a key and very important issue with `vue-easy-sync` I think there should be an added focus on two key parts to the syncing process:

SSR as a first class citizen
persistence

**Comment by mesqueeb** [comment]
@nothingismagick 
thanks for your comment and nice to see you here!

### persistence
for offline persistence we'll have to think where to best integrate that logic.
- vuex has vuex-persist
- firestore has automatic persistence in their SDK as well

I think we should make sure that our lib works well with those options.
we can also try and integrate our own persistence, but I think that might be too tricky 😅 and perhaps not necessary if we have these current solutions?

However, the user should also be able to choose "file storage" as local store option, via a plugin. this is useful for Cordova/Capacitor apps who need real persistence.

I still have to think how we can have an architecture that allows for 2 local stores. Eg. vuex + file storage. 

### SSR
I don't have experience with SSR, I'd love your input and advice once we start writing the tests & implementation!! 😍

**Comment by nothingismagick** [comment]
yeah - i think file and blob would be important from the perspective of tauri apps that i envision wanting to consume this as well.

**Comment by mesqueeb** [comment]
### Vue Sync core design:
This is a **bird's eye overview** of the architecture of "Vuex Easy Firestore 1.0" in comparison to "Vue Sync (VEF 2.0)"

[Vuex Easy Firestore & Vue Sync architecture.pdf](https://github.com/vue-sync/rfcs/files/4419398/Vuex.Easy.Firestore.Vue.Sync.architecture.pdf)

### Main improvements over 1.0:
The main improvements are being framework agnostic. While Vue Sync will launch with the Vuex & Firestore plugins, Vue Sync allows for many many more plugins to be created by the community!

### Current status:
I'm almost done with the first implementation of the core library. Soon will begin on the Vuex & Firestore plugins.

**Comment by mesqueeb** [comment]
Started writing the documentation for `core`:
https://hackmd.io/@2UCAnGkfS16bRnuAKwDapQ/vue-sync-docs

The syntax I wrote down for reading & writing data is when working directly with the "module instances" exposed by the core library. This syntax can be substituted for Vuex actions when the Vuex plugin is used.
